url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/utility/Compare/DataOfPointParser.java

```diff
@@ -0,0 +1,96 @@
+package utility.Compare;
+
+import java.io.DataInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.util.Scanner;
+
+import jdbm.PrimaryTreeMap;
+import jdbm.RecordManager;
+import jdbm.RecordManagerFactory;
+
+import utility.Compare.DataOfPoint;
+
+/**
+ * @author qchen
+ *
+ */
+public class DataOfPointParser {
+
+	public static String filein = "tmp/output";
+	public static String filename = "database/PointsData.NE";
+	public RecordManager recmanOfPoint;
+	public PrimaryTreeMap<Long, byte[]> btOfPoint;
+	
+	public void loadData() throws IOException{
+		try {
+			File files = new File(filein);
+			if(!files.exists()){
+				throw new FileNotFoundException();
+			}
+			if(files.isDirectory()){
+				File[] file = files.listFiles();
+				for(int i = 0 ; i < file.length; i++){
+					System.out.println(file[i].toString());
+					DataInputStream dis = new DataInputStream(new FileInputStream(file[i]));
+					try {
+						int lines = 0;
+						while(dis.available() != 0){
+							lines ++;
+							dis.readByte();
+							int id = dis.readInt();
+							DataOfPoint dataOfPoint = new DataOfPoint();
+							dataOfPoint.load(dis);
+							dis.readByte();
+							btOfPoint.put((long) id, dataOfPoint.writeToBytes());
+							if(lines % 100 == 0)recmanOfPoint.commit();
+						}
+						recmanOfPoint.commit();
+						dis.close();
+					} catch (IOException e) {
+						// TODO Auto-generated catch block
+						e.printStackTrace();
+					}
+				}
+			}
+			
+		} catch (FileNotFoundException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		}
+	}
+	
+	public DataOfPointParser(boolean needLoad) throws IOException{
+		recmanOfPoint = RecordManagerFactory.createRecordManager(filename);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code assigns a resource that might leak to a class field. This can cause a resource leak that slows down or crashes your system. For example, when DataOfPointParser.loadData throws an exception, and an if condition at line 68 evaluates to true, the *DataOfPointParser* object isn’t constructed so you can’t close this resource afterward.

**Fix**
Ensure that if an exception is thrown, the *recmanOfPoint* resource is closed before the constructor is exited.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).